### ISSUE TITLE
--nampace-package flag reveals more type errors but causes other problems

### ISSUE THREAD

**Issue by nbraun-wolf** [python/mypy#11260]
**Bug Report**

When the `--namespace-package` flag isn't set. Some errors go undetected. When the flag is set, the type errors are detected, but `mypy` has issues with the workspace structure and detected files double.

**To Reproduce**

First set up the project, https://github.com/nbraun-wolf/mypy-namespace-issue.

```bash
git clone git@github.com:nbraun-wolf/mypy-namespace-issue.git
export PIPENV_VENV_IN_PROJECT=1
pipenv install --ignore-pipfile
pipenv shell
export PYTHONPATH=$PWD
```

Next, lint the workspace. It passes all checks, but some problems go actually undetected.

```bash
$ mypy .
Success: no issues found in 2 source files
```

Enabling `namespace-package` is supposed to show the undetected errors, but comes with an issue of its own.

```bash
$ mypy --namespace-package .
setup.cfg: error: Source file found twice under different module names: "config" and "organization.namespace.package.core.config"
Found 1 error in 1 file (errors prevented further checking)
```

It can be more or less fixed by providing the `package` flag. With the below command, the issues are revealed.

```bash
$ mypy --namespace-package --package organization.namespace.package
organization/namespace/package/main.py:4: error: Argument "hostname" to "Client" has incompatible type "Optional[str]"; expected "str"
organization/namespace/package/main.py:4: error: Argument "port" to "Client" has incompatible type "Union[str, int]"; expected "int"
Found 2 errors in 1 file (checked 4 source files)
```

However, this isn't ideal for 2 reasons:

1. It appears to still detect the files double as we can see on its output reporting 4 files have been checked.
2. The root folder is not included in linting, which is not good in case we have other python files like or packages.

**Your Environment**

python: version: 3.8
mypy version: 0.910
windows 10 wls2 ubuntu20.04

**Comment by hauntsaninja** [comment]
Try setting `--explicit-package-bases`. See https://mypy.readthedocs.io/en/stable/running_mypy.html#mapping-file-paths-to-modules for more details on how mypy associates module names to files. #11270 might also help a little bit.

**Comment by nbraun-wolf** [comment]
Ok cool, that fixed the double file issue. Thank you :)

Now remains the question why it shows more, arguably better type hints.

No errors detected with 

```bash
$ mypy  .
Success: no issues found in 2 source files
```

But using the new modified command shows that the args passed to the function may cause problems.

```bash
$ mypy --explicit-package-bases --namespace-package  .
organization/namespace/package/main.py:4: error: Argument "hostname" to "Client" has incompatible type "Optional[str]"; expected "str"
organization/namespace/package/main.py:4: error: Argument "port" to "Client" has incompatible type "Union[str, int]"; expected "int"
Found 2 errors in 1 file (checked 2 source files)
```

Although, I just tested and adding `__init__.py` to all directories shows the same error.  

I feel like, this didn't come up last time when I tried the init files with regard to the other issue. Anyway, I guess this can be closed then, if this is the intended behaviour.

**Comment by hauntsaninja** [comment]
Probably because `ignore_missing_imports = True`: when it gets the module names wrong it can't find the imports and just replaces them with Any.
